ios/NotifeeCore/NotifeeCore.m

```diff
@@ -735,10 +735,24 @@ + (void)getInitialNotification:(notifeeMethodNSDictionaryBlock)block {
 
 + (void)setBadgeCount:(NSInteger)count withBlock:(notifeeMethodVoidBlock)block {
   if (![NotifeeCoreUtil isAppExtension]) {
-    // If count is 0, set to -1 instead to avoid notifications in tray being cleared
-    NSInteger newCount = count == 0 ? -1 : count;
-    UIApplication *application = (UIApplication *)[NotifeeCoreUtil notifeeUIApplication];
-    [application setApplicationIconBadgeNumber:newCount];
+    if (@available(iOS 16.0, macOS 10.13, macCatalyst 16.0, tvOS 16.0, visionOS 1.0, *)) {
+      UNUserNotificationCenter *center = [UNUserNotificationCenter currentNotificationCenter];
+      [center setBadgeCount:count
+          withCompletionHandler:^(NSError *error) {
+            if (error) {
+              NSLog(@"NotifeeCore: Could not setBadgeCount: %@", error);
+              block(error);
+            } else {
+              block(nil);
+            }
+          }];
+      return;
+    } else {
+      // If count is 0, set to -1 instead to avoid notifications in tray being cleared
+      NSInteger newCount = count == 0 ? 0 : count;  //-1 : count;
```

Comment:
was this change intentional?

Reply:
The comment 1 line above is now wrong but based on the commit description it would appear to be correct 🤔

Reply:
first goal was just to get the test passing, I still have final review + some cleanup to do including that comment yes

For those following along at home, what Apple did with the icon badging APIs was effectively fix a mistake

Previous situation was:

- if you wanted to set the badge to zero, you could do so, but setting it to zero was mingled with clearing the notifications from the notification center. This worked for some app's desired UX but not for others
- If you wanted to set to zero but *not* clear notifications you could use an undocumented -1 for the badge count to do so. The badge would go to zero but notifications stayed in the notification center. This was objectively ridiculous though

So, with iOS16 they added a new notification center setBadgeCount API which we had not adopted yet. With this they were explicitly trying to separate the actions of "clearing notifications" and "setting badge count to zero", when you setBadgeCount(0) the notifications stay but the badge goes away. This is what we really want.

Before now it apparently did not matter that we had not adopted the new API, but with iOS 18 Apple was extra awesome, and now if you continue using the old undocumented API of setting it to -1 to clear badge but leave notifications, it no longer actually cleared the badge. And if you call the new API with -1 it also doesn't work, you should use the real zero

This was one of the new entropic test failure that I saw.

So, I adopted use of the new API if it the device supports it, and send in a zero to clear the badge. This works in practice but now the code does need to be re-self-reviewed and cleaned up :-)

❤️  Thanks Apple ❤️

Reply:
I do appreciate any/all review though - I had a good explanation (and some plans still to clean this one up) but don't we all make mistakes all the time? I certainly do. If you see anything else please do flag it !